src/compiler/checker.ts

```diff
@@ -2064,7 +2064,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     /** Key is "/path/to/a.ts|/path/to/b.ts". */
     var amalgamatedDuplicates: Map<string, DuplicateInfoForFiles> | undefined;
     var reverseMappedCache = new Map<string, Type | undefined>();
-    var inInferTypeForHomomorphicMappedType = false;

```

Comment:
the flag was "global" and thus prevented recursing here just for any type even if `source` was different or if we'd encounter different `target`s